{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Exploring MAJIQ/VOILA outputs\n",
    "Exploring MAJIQ/VOILA outputs to understand the format and how to use them"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "alta = \"/home/bia/LandscapeSplicingGrasses/SplicingLandscapeGrasses/merging_outputs/data/alt3prime.tsv\"\n",
    "p_alta = \"/home/bia/LandscapeSplicingGrasses/SplicingLandscapeGrasses/merging_outputs/data/p_alt3prime.tsv\" # putative\n",
    "altd = \"/home/bia/LandscapeSplicingGrasses/SplicingLandscapeGrasses/merging_outputs/data/alt5prime.tsv\"\n",
    "p_altd = \"/home/bia/LandscapeSplicingGrasses/SplicingLandscapeGrasses/merging_outputs/data/p_alt5prime.tsv\"\n",
    "altad = \"/home/bia/LandscapeSplicingGrasses/SplicingLandscapeGrasses/merging_outputs/data/alt3and5prime.tsv\"\n",
    "p_altad = \"/home/bia/LandscapeSplicingGrasses/SplicingLandscapeGrasses/merging_outputs/data/p_alt3and5prime.tsv\"\n",
    "alt_first_exon = \"/home/bia/LandscapeSplicingGrasses/SplicingLandscapeGrasses/merging_outputs/data/alternate_first_exon.tsv\"\n",
    "p_alt_first_exon = \"/home/bia/LandscapeSplicingGrasses/SplicingLandscapeGrasses/merging_outputs/data/p_alternate_first_exon.tsv\"\n",
    "p_alt_last_exon = \"/home/bia/LandscapeSplicingGrasses/SplicingLandscapeGrasses/merging_outputs/data/p_alternate_last_exon.tsv\"\n",
    "alt_last_exon = \"/home/bia/LandscapeSplicingGrasses/SplicingLandscapeGrasses/merging_outputs/data/alternate_last_exon.tsv\"\n",
    "alt_intron = \"/home/bia/LandscapeSplicingGrasses/SplicingLandscapeGrasses/merging_outputs/data/alternative_intron.tsv\"\n",
    "cassette = \"/home/bia/LandscapeSplicingGrasses/SplicingLandscapeGrasses/merging_outputs/data/cassette.tsv\"\n",
    "tandem_cassette = \"/home/bia/LandscapeSplicingGrasses/SplicingLandscapeGrasses/merging_outputs/data/tandem_cassette.tsv\"\n",
    "multi_exon_spanning = \"/home/bia/LandscapeSplicingGrasses/SplicingLandscapeGrasses/merging_outputs/data/multi_exon_spanning.tsv\"\n",
    "mutually_exclusive = \"/home/bia/LandscapeSplicingGrasses/SplicingLandscapeGrasses/merging_outputs/data/mutually_exclusive.tsv\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ALT3/ALTA\n",
    "In the context of RNA splicing, an alternative acceptor (also referred to as alternative 3' splice site) is a mechanism of alternative splicing where different 3' splice sites (acceptor sites) are selected during the processing of a pre-mRNA transcript. \n",
    "\n",
    "- When 'spliced_with' is E2 and 'strand' is \"+\", 'reference_exon_coord_start' should be less than 'spliced_with_coord_start'.\n",
    "\n",
    "- When 'strand' is \"-\", the opposite should be true for E2: 'reference_exon_coord_start' should be greater than 'spliced_with_coord_start'.\n",
    "\n",
    "- For E1, everything is the opposite. So, if E1 and \"+\", reference start is greater; if E1 and \"-\", reference start is less.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "alta_df = pd.read_csv(alta, sep=\"\\t\", header = 0, comment=\"#\")\n",
    "\n",
    "# split coordinates columns in start and end\n",
    "alta_df[[\"reference_exon_coord_start\", \"reference_exon_coord_end\"]] = alta_df[\"reference_exon_coord\"].str.split(\"-\", expand=True)\n",
    "alta_df[[\"spliced_with_coord_start\", \"spliced_with_coord_end\"]] = alta_df[\"spliced_with_coord\"].str.split(\"-\", expand=True)\n",
    "alta_df[[\"junction_coord_start\", \"junction_coord_end\"]] = alta_df[\"junction_coord\"].str.split(\"-\", expand=True)\n",
    "\n",
    "# convert start and end to int\n",
    "alta_df[\"reference_exon_coord_start\"] = alta_df[\"reference_exon_coord_start\"].astype(int)\n",
    "alta_df[\"reference_exon_coord_end\"] = alta_df[\"reference_exon_coord_end\"].astype(int)\n",
    "alta_df[\"spliced_with_coord_start\"] = alta_df[\"spliced_with_coord_start\"].astype(int)\n",
    "alta_df[\"spliced_with_coord_end\"] = alta_df[\"spliced_with_coord_end\"].astype(int)\n",
    "alta_df[\"junction_coord_start\"] = alta_df[\"junction_coord_start\"].astype(int)\n",
    "alta_df[\"junction_coord_end\"] = alta_df[\"junction_coord_end\"].astype(int)\n",
    "\n",
    "# print(alta_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Linhas que violam a teoria:\n",
      "Empty DataFrame\n",
      "Columns: [module_id, gene_id, gene_name, seqid, strand, lsv_id, event_id, complex, denovo, reference_exon_coord, spliced_with, spliced_with_coord, junction_name, junction_coord, event_size, event_non_changing, event_changing, junction_changing, SRR11684912_median_psi, SRR11684912_var_psi, reference_exon_coord_start, reference_exon_coord_end, spliced_with_coord_start, spliced_with_coord_end, junction_coord_start, junction_coord_end, condicao_valida]\n",
      "Index: []\n",
      "\n",
      "[0 rows x 27 columns]\n",
      "E2    2876\n",
      "E1      98\n",
      "Name: spliced_with, dtype: int64 True    2974\n",
      "Name: condicao_valida, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "condicao_e2_positivo = (alta_df[\"spliced_with\"] == \"E2\") & (alta_df[\"strand\"] == \"+\") & (alta_df[\"reference_exon_coord_start\"] < alta_df[\"spliced_with_coord_start\"])\n",
    "condicao_e2_negativo = (alta_df[\"spliced_with\"] == \"E2\") & (alta_df[\"strand\"] == \"-\") & (alta_df[\"reference_exon_coord_start\"] > alta_df[\"spliced_with_coord_start\"])\n",
    "condicao_e1_positivo = (alta_df[\"spliced_with\"] == \"E1\") & (alta_df[\"strand\"] == \"+\") & (alta_df[\"reference_exon_coord_start\"] > alta_df[\"spliced_with_coord_start\"])\n",
    "condicao_e1_negativo = (alta_df[\"spliced_with\"] == \"E1\") & (alta_df[\"strand\"] == \"-\") & (alta_df[\"reference_exon_coord_start\"] < alta_df[\"spliced_with_coord_start\"])\n",
    "\n",
    "alta_df[\"condicao_valida\"] = condicao_e2_positivo | condicao_e2_negativo | condicao_e1_positivo | condicao_e1_negativo\n",
    "linhas_invalidas = alta_df[~alta_df[\"condicao_valida\"]]\n",
    "print(\"Linhas que violam a teoria:\")\n",
    "print(linhas_invalidas)\n",
    "print(alta_df[\"spliced_with\"].value_counts(), alta_df[\"condicao_valida\"].value_counts())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ALT5/ALTD\n",
    "In RNA splicing, an alternative donor (or alternative 5' splice site) refers to a mechanism where different 5' splice sites (donor sites) are selected during pre-mRNA processing.\n",
    "\n",
    "- When 'spliced_with' is E2 and 'strand' is \"+\", 'reference_exon_coord_start' should be less than 'spliced_with_coord_start'.\n",
    "\n",
    "- When 'strand' is \"-\", the opposite should be true for E2: 'reference_exon_coord_start' should be greater than 'spliced_with_coord_start'.\n",
    "\n",
    "- For E1, everything is the opposite. So, if E1 and \"+\", reference start is greater; if E1 and \"-\", reference start is less.\n",
    "\n",
    "- ! Unknown coordinates are marked as -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                   module_id                  gene_id  gene_name seqid strand  \\\n",
      "0  AT3G29160.Araport11.447_2  AT3G29160.Araport11.447  AT3G29160  Chr3      -   \n",
      "1  AT3G29160.Araport11.447_2  AT3G29160.Araport11.447  AT3G29160  Chr3      -   \n",
      "2  AT1G06010.Araport11.447_1  AT1G06010.Araport11.447  AT1G06010  Chr1      +   \n",
      "3  AT1G06010.Araport11.447_1  AT1G06010.Araport11.447  AT1G06010  Chr1      +   \n",
      "4  AT2G03810.Araport11.447_1  AT2G03810.Araport11.447  AT2G03810  Chr2      +   \n",
      "\n",
      "                                        lsv_id  \\\n",
      "0  AT3G29160.Araport11.447:s:11131004-11131188   \n",
      "1  AT3G29160.Araport11.447:s:11131004-11131188   \n",
      "2    AT1G06010.Araport11.447:s:1823344-1823410   \n",
      "3    AT1G06010.Araport11.447:s:1823344-1823410   \n",
      "4    AT2G03810.Araport11.447:t:1162633-1162712   \n",
      "\n",
      "                         event_id  complex  denovo reference_exon_coord  ...  \\\n",
      "0  AT3G29160.Araport11.447_2_A5_1    False   False    11131004-11131188  ...   \n",
      "1  AT3G29160.Araport11.447_2_A5_1    False    True    11131004-11131188  ...   \n",
      "2  AT1G06010.Araport11.447_1_A5_1    False   False      1823344-1823410  ...   \n",
      "3  AT1G06010.Araport11.447_1_A5_1    False    True      1823344-1823410  ...   \n",
      "4  AT2G03810.Araport11.447_1_A5_1     True   False      1162633-1162712  ...   \n",
      "\n",
      "  event_changing junction_changing SRR11684912_median_psi SRR11684912_var_psi  \\\n",
      "0            NaN               NaN                0.91160            0.002315   \n",
      "1            NaN               NaN                0.08844            0.002315   \n",
      "2            NaN               NaN                0.94290            0.001137   \n",
      "3            NaN               NaN                0.05711            0.001137   \n",
      "4            NaN               NaN                0.36540            0.011040   \n",
      "\n",
      "   reference_exon_coord_start  reference_exon_coord_end  \\\n",
      "0                    11131004                  11131188   \n",
      "1                    11131004                  11131188   \n",
      "2                     1823344                   1823410   \n",
      "3                     1823344                   1823410   \n",
      "4                     1162633                   1162712   \n",
      "\n",
      "   spliced_with_coord_start  spliced_with_coord_end  junction_coord_start  \\\n",
      "0                  11130727                11130909              11130909   \n",
      "1                  11130727                11130909              11130909   \n",
      "2                   1823498                 1824236               1823410   \n",
      "3                   1823498                 1824236               1823385   \n",
      "4                   1161526                 1161729               1161729   \n",
      "\n",
      "   junction_coord_end  \n",
      "0            11131004  \n",
      "1            11131127  \n",
      "2             1823498  \n",
      "3             1823498  \n",
      "4             1162633  \n",
      "\n",
      "[5 rows x 26 columns]\n"
     ]
    }
   ],
   "source": [
    "altd_df = pd.read_csv(altd, sep=\"\\t\", header=0, comment=\"#\")\n",
    "\n",
    "def split_coord(coord):\n",
    "    if coord.startswith(\"-1-\"):\n",
    "        return \"-1\", coord.split(\"-1-\", 1)[1]\n",
    "    elif coord.endswith(\"--1\"):\n",
    "        return coord.split(\"--1\", 1)[0], \"-1\"\n",
    "    else:\n",
    "        return coord.split(\"-\", 1)\n",
    "\n",
    "# split coordinates columns in start and end\n",
    "altd_df[[\"reference_exon_coord_start\", \"reference_exon_coord_end\"]] = altd_df[\"reference_exon_coord\"].str.split(\"-\", expand=True)\n",
    "altd_df[[\"spliced_with_coord_start\", \"spliced_with_coord_end\"]] = altd_df[\"spliced_with_coord\"].apply(split_coord).tolist()\n",
    "altd_df[[\"junction_coord_start\", \"junction_coord_end\"]] = altd_df[\"junction_coord\"].str.split(\"-\", expand=True)\n",
    "\n",
    "# convert start and end to int\n",
    "altd_df[\"reference_exon_coord_start\"] = altd_df[\"reference_exon_coord_start\"].astype(int)\n",
    "altd_df[\"reference_exon_coord_end\"] = altd_df[\"reference_exon_coord_end\"].astype(int)\n",
    "altd_df[\"spliced_with_coord_start\"] = altd_df[\"spliced_with_coord_start\"].astype(int)\n",
    "altd_df[\"spliced_with_coord_end\"] = altd_df[\"spliced_with_coord_end\"].astype(int)\n",
    "altd_df[\"junction_coord_start\"] = altd_df[\"junction_coord_start\"].astype(int)\n",
    "altd_df[\"junction_coord_end\"] = altd_df[\"junction_coord_end\"].astype(int)\n",
    "\n",
    "print(altd_df.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Linhas que violam a teoria:\n",
      "Empty DataFrame\n",
      "Columns: [module_id, gene_id, gene_name, seqid, strand, lsv_id, event_id, complex, denovo, reference_exon_coord, spliced_with, spliced_with_coord, junction_name, junction_coord, event_size, event_non_changing, event_changing, junction_changing, SRR11684912_median_psi, SRR11684912_var_psi, reference_exon_coord_start, reference_exon_coord_end, spliced_with_coord_start, spliced_with_coord_end, junction_coord_start, junction_coord_end, condicao_valida]\n",
      "Index: []\n",
      "\n",
      "[0 rows x 27 columns]\n",
      "E2    1296\n",
      "E1     108\n",
      "Name: spliced_with, dtype: int64 True    1404\n",
      "Name: condicao_valida, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "condicao_e2_positivo = (altd_df[\"spliced_with\"] == \"E2\") & (altd_df[\"strand\"] == \"+\") & (altd_df[\"reference_exon_coord_start\"] < altd_df[\"spliced_with_coord_start\"])\n",
    "condicao_e2_negativo = (altd_df[\"spliced_with\"] == \"E2\") & (altd_df[\"strand\"] == \"-\") & (altd_df[\"reference_exon_coord_start\"] > altd_df[\"spliced_with_coord_start\"])\n",
    "condicao_e1_positivo = (altd_df[\"spliced_with\"] == \"E1\") & (altd_df[\"strand\"] == \"+\") & (altd_df[\"reference_exon_coord_start\"] > altd_df[\"spliced_with_coord_start\"])\n",
    "condicao_e1_negativo = (altd_df[\"spliced_with\"] == \"E1\") & (altd_df[\"strand\"] == \"-\") & (altd_df[\"reference_exon_coord_start\"] < altd_df[\"spliced_with_coord_start\"])\n",
    "\n",
    "altd_df[\"condicao_valida\"] = condicao_e2_positivo | condicao_e2_negativo | condicao_e1_positivo | condicao_e1_negativo\n",
    "linhas_invalidas = altd_df[~altd_df[\"condicao_valida\"]]\n",
    "print(\"Linhas que violam a teoria:\")\n",
    "print(linhas_invalidas)\n",
    "print(altd_df[\"spliced_with\"].value_counts(), altd_df[\"condicao_valida\"].value_counts())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ALTA and ALTD \n",
    "\n",
    "- When 'spliced_with' is E2 and 'strand' is \"+\", 'reference_exon_coord_start' should be less than 'spliced_with_coord_start'.\n",
    "\n",
    "- When 'strand' is \"-\", the opposite should be true for E2: 'reference_exon_coord_start' should be greater than 'spliced_with_coord_start'.\n",
    "\n",
    "- For E1, everything is the opposite. So, if E1 and \"+\", reference start is greater; if E1 and \"-\", reference start is less.\n",
    "\n",
    "- lines with NaN values in lsv_id column are duplicates and were removed \n",
    "\n",
    "- reference_exon_coord and spliced_with_coord are the same for ALTA and ALTD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "altad_df = pd.read_csv(altad, sep=\"\\t\", header=0, comment=\"#\")\n",
    "\n",
    "def check_inverted_duplicates(df):\n",
    "    for i in range(0, len(df), 4):  # Processa grupos de 4 linhas\n",
    "        group1 = df.iloc[i:i+2]  # Linhas 1 e 2\n",
    "        group2 = df.iloc[i+2:i+4]  # Linhas 3 e 4\n",
    "\n",
    "        if not (group1[\"reference_exon_coord\"].nunique() == 1 and\n",
    "                group1[\"spliced_with_coord\"].nunique() == 1):\n",
    "            print(f\"Erro de consistência no grupo {i//4 + 1}:\")\n",
    "            print(group1)\n",
    "            continue \n",
    "\n",
    "        # Compara linha 1 com 3 e linha 2 com 4\n",
    "        for j in range(2):\n",
    "            row1 = group1.iloc[j]\n",
    "            row2 = group2.iloc[j]\n",
    "\n",
    "            # Verifica se as colunas estão invertidas\n",
    "            if (row1[\"reference_exon_coord\"] == row2[\"spliced_with_coord\"] and\n",
    "                row1[\"spliced_with_coord\"] == row2[\"reference_exon_coord\"]):\n",
    "                # print(f\"Linha {i + j} e {i + 2 + j} são duplicatas invertidas.\")\n",
    "                continue\n",
    "            else:\n",
    "                print(f\"Linha {i + j} e {i + 2 + j} NÃO são duplicatas invertidas.\")\n",
    "\n",
    "\n",
    "# Aplicar a função ao DataFrame\n",
    "check_inverted_duplicates(altad_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                   module_id                  gene_id  gene_name seqid strand  \\\n",
      "0  AT5G15540.Araport11.447_1  AT5G15540.Araport11.447  AT5G15540  Chr5      -   \n",
      "1  AT5G15540.Araport11.447_1  AT5G15540.Araport11.447  AT5G15540  Chr5      -   \n",
      "4  AT3G54500.Araport11.447_1  AT3G54500.Araport11.447  AT3G54500  Chr3      -   \n",
      "5  AT3G54500.Araport11.447_1  AT3G54500.Araport11.447  AT3G54500  Chr3      -   \n",
      "8  AT3G54500.Araport11.447_2  AT3G54500.Araport11.447  AT3G54500  Chr3      -   \n",
      "\n",
      "                                        lsv_id  \\\n",
      "0    AT5G15540.Araport11.447:s:5056537-5056646   \n",
      "1    AT5G15540.Araport11.447:s:5056537-5056646   \n",
      "4  AT3G54500.Araport11.447:s:20178940-20179070   \n",
      "5  AT3G54500.Araport11.447:s:20178940-20179070   \n",
      "8  AT3G54500.Araport11.447:s:20178123-20178491   \n",
      "\n",
      "                           event_id  complex  denovo reference_exon_coord  \\\n",
      "0  AT5G15540.Araport11.447_1_A3A5_1    False   False      5056537-5056646   \n",
      "1  AT5G15540.Araport11.447_1_A3A5_1    False   False      5056537-5056646   \n",
      "4  AT3G54500.Araport11.447_1_A3A5_1    False   False    20178940-20179070   \n",
      "5  AT3G54500.Araport11.447_1_A3A5_1    False    True    20178940-20179070   \n",
      "8  AT3G54500.Araport11.447_2_A3A5_1     True   False    20178123-20178491   \n",
      "\n",
      "   ... event_changing junction_changing SRR11684912_median_psi  \\\n",
      "0  ...          False               NaN                 0.1108   \n",
      "1  ...          False               NaN                 0.8892   \n",
      "4  ...          False               NaN                 0.9420   \n",
      "5  ...          False               NaN                 0.0580   \n",
      "8  ...          False               NaN                 0.5675   \n",
      "\n",
      "  SRR11684912_var_psi reference_exon_coord_start reference_exon_coord_end  \\\n",
      "0            0.004905                    5056537                  5056646   \n",
      "1            0.004905                    5056537                  5056646   \n",
      "4            0.000596                   20178940                 20179070   \n",
      "5            0.000596                   20178940                 20179070   \n",
      "8            0.001522                   20178123                 20178491   \n",
      "\n",
      "   spliced_with_coord_start  spliced_with_coord_end  junction_coord_start  \\\n",
      "0                   5055962                 5056217               5056217   \n",
      "1                   5055962                 5056217               5056209   \n",
      "4                  20178714                20178804              20178804   \n",
      "5                  20178714                20178804              20178800   \n",
      "8                  20177935                20178064              20178064   \n",
      "\n",
      "   junction_coord_end  \n",
      "0             5056551  \n",
      "1             5056537  \n",
      "4            20178941  \n",
      "5            20178940  \n",
      "8            20178158  \n",
      "\n",
      "[5 rows x 25 columns]\n"
     ]
    }
   ],
   "source": [
    "altad_df = altad_df.dropna(subset=[\"lsv_id\"])\n",
    "\n",
    "# split coordinates columns in start and end\n",
    "altad_df[[\"reference_exon_coord_start\", \"reference_exon_coord_end\"]] = altad_df[\"reference_exon_coord\"].str.split(\"-\", expand=True)\n",
    "altad_df[[\"spliced_with_coord_start\", \"spliced_with_coord_end\"]] = altad_df[\"spliced_with_coord\"].str.split(\"-\", expand=True)\n",
    "altad_df[[\"junction_coord_start\", \"junction_coord_end\"]] = altad_df[\"junction_coord\"].str.split(\"-\", expand=True)\n",
    "\n",
    "# convert start and end to int\n",
    "altad_df[\"reference_exon_coord_start\"] = altad_df[\"reference_exon_coord_start\"].astype(int)\n",
    "altad_df[\"reference_exon_coord_end\"] = altad_df[\"reference_exon_coord_end\"].astype(int)\n",
    "altad_df[\"spliced_with_coord_start\"] = altad_df[\"spliced_with_coord_start\"].astype(int)\n",
    "altad_df[\"spliced_with_coord_end\"] = altad_df[\"spliced_with_coord_end\"].astype(int)\n",
    "altad_df[\"junction_coord_start\"] = altad_df[\"junction_coord_start\"].astype(int)\n",
    "altad_df[\"junction_coord_end\"] = altad_df[\"junction_coord_end\"].astype(int)\n",
    "\n",
    "print(altad_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Linhas que violam a teoria:\n",
      "Empty DataFrame\n",
      "Columns: [module_id, gene_id, gene_name, seqid, strand, lsv_id, event_id, complex, denovo, reference_exon_coord, spliced_with, spliced_with_coord, junction_name, junction_coord, event_size, event_non_changing, event_changing, junction_changing, SRR11684912_median_psi, SRR11684912_var_psi, reference_exon_coord_start, reference_exon_coord_end, spliced_with_coord_start, spliced_with_coord_end, junction_coord_start, junction_coord_end, condicao_valida]\n",
      "Index: []\n",
      "\n",
      "[0 rows x 27 columns]\n",
      "E2    192\n",
      "E1     14\n",
      "Name: spliced_with, dtype: int64 True    206\n",
      "Name: condicao_valida, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "condicao_e2_positivo = (altad_df[\"spliced_with\"] == \"E2\") & (altad_df[\"strand\"] == \"+\") & (altad_df[\"reference_exon_coord_start\"] < altad_df[\"spliced_with_coord_start\"])\n",
    "condicao_e2_negativo = (altad_df[\"spliced_with\"] == \"E2\") & (altad_df[\"strand\"] == \"-\") & (altad_df[\"reference_exon_coord_start\"] > altad_df[\"spliced_with_coord_start\"])\n",
    "condicao_e1_positivo = (altad_df[\"spliced_with\"] == \"E1\") & (altad_df[\"strand\"] == \"+\") & (altad_df[\"reference_exon_coord_start\"] > altad_df[\"spliced_with_coord_start\"])\n",
    "condicao_e1_negativo = (altad_df[\"spliced_with\"] == \"E1\") & (altad_df[\"strand\"] == \"-\") & (altad_df[\"reference_exon_coord_start\"] < altad_df[\"spliced_with_coord_start\"])\n",
    "\n",
    "altad_df[\"condicao_valida\"] = condicao_e2_positivo | condicao_e2_negativo | condicao_e1_positivo | condicao_e1_negativo\n",
    "linhas_invalidas = altd_df[~altd_df[\"condicao_valida\"]]\n",
    "print(\"Linhas que violam a teoria:\")\n",
    "print(linhas_invalidas)\n",
    "print(altad_df[\"spliced_with\"].value_counts(), altad_df[\"condicao_valida\"].value_counts())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Alternate first exon\n",
    "In RNA splicing, alternative first exon refers to a mechanism where different initial exons (first exons) are used to start the transcription of a gene. This occurs when a gene has multiple promoters or transcription start sites, each associated with a unique first exon\n",
    "\n",
    "- When 'spliced_with' is E2 and 'strand' is \"+\", 'reference_exon_coord_start' should be less than 'spliced_with_coord_start'.\n",
    "\n",
    "- When 'strand' is \"-\", the opposite should be true for E2: 'reference_exon_coord_start' should be greater than 'spliced_with_coord_start'.\n",
    "\n",
    "- For E1, everything is the opposite. So, if E1 and \"+\", reference start is greater; if E1 and \"-\", reference start is less.\n",
    "\n",
    "- Spliced with is always A\n",
    "\n",
    "- ! Unknown coordinates are marked as -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                   module_id                  gene_id  gene_name seqid strand  \\\n",
      "0  AT3G29160.Araport11.447_1  AT3G29160.Araport11.447  AT3G29160  Chr3      -   \n",
      "1  AT3G29160.Araport11.447_1  AT3G29160.Araport11.447  AT3G29160  Chr3      -   \n",
      "2  AT2G03620.Araport11.447_1  AT2G03620.Araport11.447  AT2G03620  Chr2      -   \n",
      "3  AT2G03620.Araport11.447_1  AT2G03620.Araport11.447  AT2G03620  Chr2      -   \n",
      "4  AT2G03810.Araport11.447_1  AT2G03810.Araport11.447  AT2G03810  Chr2      +   \n",
      "\n",
      "                                        lsv_id  \\\n",
      "0  AT3G29160.Araport11.447:t:11131318-11131513   \n",
      "1  AT3G29160.Araport11.447:t:11131318-11131513   \n",
      "2    AT2G03620.Araport11.447:t:1101727-1102292   \n",
      "3    AT2G03620.Araport11.447:t:1101727-1102292   \n",
      "4    AT2G03810.Araport11.447:t:1162633-1162712   \n",
      "\n",
      "                          event_id  complex  denovo reference_exon_coord  ...  \\\n",
      "0  AT3G29160.Araport11.447_1_afe_1    False   False    11131318-11131513  ...   \n",
      "1  AT3G29160.Araport11.447_1_afe_1    False   False    11131318-11131513  ...   \n",
      "2  AT2G03620.Araport11.447_1_afe_1    False   False      1101727-1102292  ...   \n",
      "3  AT2G03620.Araport11.447_1_afe_1    False   False      1101727-1102292  ...   \n",
      "4  AT2G03810.Araport11.447_1_afe_1     True   False      1162633-1162712  ...   \n",
      "\n",
      "  event_changing junction_changing SRR11684912_median_psi SRR11684912_var_psi  \\\n",
      "0            NaN               NaN                 0.7543            0.003609   \n",
      "1            NaN               NaN                 0.2457            0.003609   \n",
      "2            NaN               NaN                 0.8141            0.021990   \n",
      "3            NaN               NaN                 0.1859            0.021990   \n",
      "4            NaN               NaN                 0.5074            0.011930   \n",
      "\n",
      "   reference_exon_coord_start  reference_exon_coord_end  \\\n",
      "0                    11131318                  11131513   \n",
      "1                    11131318                  11131513   \n",
      "2                     1101727                   1102292   \n",
      "3                     1101727                   1102292   \n",
      "4                     1162633                   1162712   \n",
      "\n",
      "   spliced_with_coord_start  spliced_with_coord_end  junction_coord_start  \\\n",
      "0                  11131809                11131883              11131513   \n",
      "1                  11131957                11132105              11131513   \n",
      "2                   1102714                 1102811               1102292   \n",
      "3                   1102967                 1103211               1102292   \n",
      "4                   1161526                 1161724               1161724   \n",
      "\n",
      "   junction_coord_end  \n",
      "0            11131809  \n",
      "1            11131957  \n",
      "2             1102714  \n",
      "3             1102967  \n",
      "4             1162633  \n",
      "\n",
      "[5 rows x 25 columns]\n"
     ]
    }
   ],
   "source": [
    "alt_first_exon_df = pd.read_csv(alt_first_exon, sep=\"\\t\", header=0, comment=\"#\")\n",
    "\n",
    "# split coordinates columns in start and end\n",
    "alt_first_exon_df[[\"reference_exon_coord_start\", \"reference_exon_coord_end\"]] = alt_first_exon_df[\"reference_exon_coord\"].str.split(\"-\", expand=True)\n",
    "alt_first_exon_df[[\"spliced_with_coord_start\", \"spliced_with_coord_end\"]] = alt_first_exon_df[\"spliced_with_coord\"].apply(split_coord).tolist()\n",
    "alt_first_exon_df[[\"junction_coord_start\", \"junction_coord_end\"]] = alt_first_exon_df[\"junction_coord\"].str.split(\"-\", expand=True)\n",
    "\n",
    "# convert start and end to int\n",
    "alt_first_exon_df[\"reference_exon_coord_start\"] = alt_first_exon_df[\"reference_exon_coord_start\"].astype(int)\n",
    "alt_first_exon_df[\"reference_exon_coord_end\"] = alt_first_exon_df[\"reference_exon_coord_end\"].astype(int)\n",
    "alt_first_exon_df[\"spliced_with_coord_start\"] = alt_first_exon_df[\"spliced_with_coord_start\"].astype(int)\n",
    "alt_first_exon_df[\"spliced_with_coord_end\"] = alt_first_exon_df[\"spliced_with_coord_end\"].astype(int)\n",
    "alt_first_exon_df[\"junction_coord_start\"] = alt_first_exon_df[\"junction_coord_start\"].astype(int)\n",
    "alt_first_exon_df[\"junction_coord_end\"] = alt_first_exon_df[\"junction_coord_end\"].astype(int)\n",
    "\n",
    "print(alt_first_exon_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Linhas que violam a teoria:\n",
      "Empty DataFrame\n",
      "Columns: [module_id, gene_id, gene_name, seqid, strand, lsv_id, event_id, complex, denovo, reference_exon_coord, spliced_with, spliced_with_coord, junction_name, junction_coord, event_non_changing, event_changing, junction_changing, SRR11684912_median_psi, SRR11684912_var_psi, reference_exon_coord_start, reference_exon_coord_end, spliced_with_coord_start, spliced_with_coord_end, junction_coord_start, junction_coord_end, condicao_valida]\n",
      "Index: []\n",
      "\n",
      "[0 rows x 26 columns]\n",
      "A    428\n",
      "Name: spliced_with, dtype: int64 True    428\n",
      "Name: condicao_valida, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "condicao_e2_positivo = (alt_first_exon_df[\"spliced_with\"] == \"A\") & (alt_first_exon_df[\"strand\"] == \"+\") & (alt_first_exon_df[\"reference_exon_coord_start\"] < alt_first_exon_df[\"spliced_with_coord_start\"])\n",
    "condicao_e2_negativo = (alt_first_exon_df[\"spliced_with\"] == \"A\") & (alt_first_exon_df[\"strand\"] == \"-\") & (alt_first_exon_df[\"reference_exon_coord_start\"] > alt_first_exon_df[\"spliced_with_coord_start\"])\n",
    "condicao_e1_positivo = (alt_first_exon_df[\"spliced_with\"] == \"A\") & (alt_first_exon_df[\"strand\"] == \"+\") & (alt_first_exon_df[\"reference_exon_coord_start\"] > alt_first_exon_df[\"spliced_with_coord_start\"])\n",
    "condicao_e1_negativo = (alt_first_exon_df[\"spliced_with\"] == \"A\") & (alt_first_exon_df[\"strand\"] == \"-\") & (alt_first_exon_df[\"reference_exon_coord_start\"] < alt_first_exon_df[\"spliced_with_coord_start\"])\n",
    "\n",
    "alt_first_exon_df[\"condicao_valida\"] = condicao_e2_positivo | condicao_e2_negativo | condicao_e1_positivo | condicao_e1_negativo\n",
    "linhas_invalidas = alt_first_exon_df[~alt_first_exon_df[\"condicao_valida\"]]\n",
    "print(\"Linhas que violam a teoria:\")\n",
    "print(linhas_invalidas)\n",
    "print(alt_first_exon_df[\"spliced_with\"].value_counts(), alt_first_exon_df[\"condicao_valida\"].value_counts())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Alternate last exon \n",
    "In RNA splicing, alternative last exon refers to a mechanism where different terminal exons (last exons) are used to conclude the transcription of a gene. This occurs when a gene has multiple polyadenylation sites or transcription termination signals, each associated with a unique last exon.\n",
    "\n",
    "- When 'spliced_with' is E2 and 'strand' is \"+\", 'reference_exon_coord_start' should be less than 'spliced_with_coord_start'.\n",
    "\n",
    "- When 'strand' is \"-\", the opposite should be true for E2: 'reference_exon_coord_start' should be greater than 'spliced_with_coord_start'.\n",
    "\n",
    "- For E1, everything is the opposite. So, if E1 and \"+\", reference start is greater; if E1 and \"-\", reference start is less.\n",
    "\n",
    "- Spliced with is always A\n",
    "\n",
    "- ! Unknown coordinates are marked as -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                   module_id                  gene_id  gene_name seqid strand  \\\n",
      "0  AT1G70850.Araport11.447_1  AT1G70850.Araport11.447  AT1G70850  Chr1      -   \n",
      "1  AT1G70850.Araport11.447_1  AT1G70850.Araport11.447  AT1G70850  Chr1      -   \n",
      "2  AT1G70850.Araport11.447_1  AT1G70850.Araport11.447  AT1G70850  Chr1      -   \n",
      "3  AT1G70850.Araport11.447_1  AT1G70850.Araport11.447  AT1G70850  Chr1      -   \n",
      "4  AT5G03560.Araport11.447_1  AT5G03560.Araport11.447  AT5G03560  Chr5      -   \n",
      "\n",
      "                                        lsv_id  \\\n",
      "0  AT1G70850.Araport11.447:s:26716540-26716901   \n",
      "1  AT1G70850.Araport11.447:s:26716540-26716901   \n",
      "2  AT1G70850.Araport11.447:s:26716540-26716901   \n",
      "3  AT1G70850.Araport11.447:s:26716540-26716901   \n",
      "4      AT5G03560.Araport11.447:s:902412-902770   \n",
      "\n",
      "                          event_id  complex  denovo reference_exon_coord  ...  \\\n",
      "0  AT1G70850.Araport11.447_1_ale_1     True    True    26716540-26716901  ...   \n",
      "1  AT1G70850.Araport11.447_1_ale_1     True    True    26716540-26716901  ...   \n",
      "2  AT1G70850.Araport11.447_1_ale_2     True   False    26716540-26716901  ...   \n",
      "3  AT1G70850.Araport11.447_1_ale_2     True    True    26716540-26716901  ...   \n",
      "4  AT5G03560.Araport11.447_1_ale_1    False   False        902412-902770  ...   \n",
      "\n",
      "  event_changing junction_changing SRR11684912_median_psi SRR11684912_var_psi  \\\n",
      "0            NaN               NaN                0.04170            0.000252   \n",
      "1            NaN               NaN                0.08408            0.000487   \n",
      "2            NaN               NaN                0.86700            0.000754   \n",
      "3            NaN               NaN                0.04170            0.000252   \n",
      "4            NaN               NaN                0.08145            0.001271   \n",
      "\n",
      "   reference_exon_coord_start  reference_exon_coord_end  \\\n",
      "0                    26716540                  26716901   \n",
      "1                    26716540                  26716901   \n",
      "2                    26716540                  26716901   \n",
      "3                    26716540                  26716901   \n",
      "4                      902412                    902770   \n",
      "\n",
      "   spliced_with_coord_start  spliced_with_coord_end  junction_coord_start  \\\n",
      "0                  26715172                26715706              26715706   \n",
      "1                        -1                26711053              26711053   \n",
      "2                  26715933                26716400              26716400   \n",
      "3                  26715172                26715706              26715706   \n",
      "4                    901425                  902235                902235   \n",
      "\n",
      "   junction_coord_end  \n",
      "0            26716540  \n",
      "1            26716540  \n",
      "2            26716540  \n",
      "3            26716540  \n",
      "4              902412  \n",
      "\n",
      "[5 rows x 25 columns]\n"
     ]
    }
   ],
   "source": [
    "alt_last_exon_df = pd.read_csv(alt_last_exon, sep=\"\\t\", header=0, comment=\"#\")\n",
    "\n",
    "# split coordinates columns in start and end\n",
    "alt_last_exon_df[[\"reference_exon_coord_start\", \"reference_exon_coord_end\"]] = alt_last_exon_df[\"reference_exon_coord\"].str.split(\"-\", expand=True)\n",
    "alt_last_exon_df[[\"spliced_with_coord_start\", \"spliced_with_coord_end\"]] = alt_last_exon_df[\"spliced_with_coord\"].apply(split_coord).tolist()\n",
    "alt_last_exon_df[[\"junction_coord_start\", \"junction_coord_end\"]] = alt_last_exon_df[\"junction_coord\"].str.split(\"-\", expand=True)\n",
    "\n",
    "# convert start and end to int\n",
    "alt_last_exon_df[\"reference_exon_coord_start\"] = alt_last_exon_df[\"reference_exon_coord_start\"].astype(int)\n",
    "alt_last_exon_df[\"reference_exon_coord_end\"] = alt_last_exon_df[\"reference_exon_coord_end\"].astype(int)\n",
    "alt_last_exon_df[\"spliced_with_coord_start\"] = alt_last_exon_df[\"spliced_with_coord_start\"].astype(int)\n",
    "alt_last_exon_df[\"spliced_with_coord_end\"] = alt_last_exon_df[\"spliced_with_coord_end\"].astype(int)\n",
    "alt_last_exon_df[\"junction_coord_start\"] = alt_last_exon_df[\"junction_coord_start\"].astype(int)\n",
    "alt_last_exon_df[\"junction_coord_end\"] = alt_last_exon_df[\"junction_coord_end\"].astype(int)\n",
    "\n",
    "print(alt_last_exon_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Linhas que violam a teoria:\n",
      "Empty DataFrame\n",
      "Columns: [module_id, gene_id, gene_name, seqid, strand, lsv_id, event_id, complex, denovo, reference_exon_coord, spliced_with, spliced_with_coord, junction_name, junction_coord, event_non_changing, event_changing, junction_changing, SRR11684912_median_psi, SRR11684912_var_psi, reference_exon_coord_start, reference_exon_coord_end, spliced_with_coord_start, spliced_with_coord_end, junction_coord_start, junction_coord_end, condicao_valida]\n",
      "Index: []\n",
      "\n",
      "[0 rows x 26 columns]\n",
      "A    138\n",
      "Name: spliced_with, dtype: int64 True    138\n",
      "Name: condicao_valida, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "condicao_e2_positivo = (alt_last_exon_df[\"spliced_with\"] == \"A\") & (alt_last_exon_df[\"strand\"] == \"+\") & (alt_last_exon_df[\"reference_exon_coord_start\"] < alt_last_exon_df[\"spliced_with_coord_start\"])\n",
    "condicao_e2_negativo = (alt_last_exon_df[\"spliced_with\"] == \"A\") & (alt_last_exon_df[\"strand\"] == \"-\") & (alt_last_exon_df[\"reference_exon_coord_start\"] > alt_last_exon_df[\"spliced_with_coord_start\"])\n",
    "condicao_e1_positivo = (alt_last_exon_df[\"spliced_with\"] == \"A\") & (alt_last_exon_df[\"strand\"] == \"+\") & (alt_last_exon_df[\"reference_exon_coord_start\"] > alt_last_exon_df[\"spliced_with_coord_start\"])\n",
    "condicao_e1_negativo = (alt_last_exon_df[\"spliced_with\"] == \"A\") & (alt_last_exon_df[\"strand\"] == \"-\") & (alt_last_exon_df[\"reference_exon_coord_start\"] < alt_last_exon_df[\"spliced_with_coord_start\"])\n",
    "\n",
    "alt_last_exon_df[\"condicao_valida\"] = condicao_e2_positivo | condicao_e2_negativo | condicao_e1_positivo | condicao_e1_negativo\n",
    "linhas_invalidas = alt_last_exon_df[~alt_last_exon_df[\"condicao_valida\"]]\n",
    "print(\"Linhas que violam a teoria:\")\n",
    "print(linhas_invalidas)\n",
    "print(alt_last_exon_df[\"spliced_with\"].value_counts(), alt_last_exon_df[\"condicao_valida\"].value_counts())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Intron retention or Alternative Intron\n",
    "!!!!!!!!!!!!!!!!\n",
    "Intron retention is a mechanism of alternative splicing where an intron, which is typically removed during the processing of pre-mRNA into mature mRNA, is retained in the final transcript. Instead of being spliced out, the intron remains within the mRNA, effectively becoming part of the coding or non-coding sequence.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0     C1_C2_intron\n",
      "1    C1_C2_spliced\n",
      "2     C1_C2_intron\n",
      "3    C1_C2_spliced\n",
      "4     C1_C2_intron\n",
      "Name: junction_name, dtype: object\n"
     ]
    }
   ],
   "source": [
    "alt_intron_df = pd.read_csv(alt_intron, sep=\"\\t\", header=0, comment=\"#\")\n",
    "\n",
    "# split coordinates columns in start and end\n",
    "alt_intron_df[[\"reference_exon_coord_start\", \"reference_exon_coord_end\"]] = alt_intron_df[\"reference_exon_coord\"].str.split(\"-\", expand=True)\n",
    "alt_intron_df[[\"spliced_with_coord_start\", \"spliced_with_coord_end\"]] = alt_intron_df[\"spliced_with_coord\"].str.split(\"-\", expand=True)\n",
    "alt_intron_df[[\"junction_coord_start\", \"junction_coord_end\"]] = alt_intron_df[\"junction_coord\"].str.split(\"-\", expand=True)\n",
    "\n",
    "# convert start and end to int\n",
    "alt_intron_df[\"reference_exon_coord_start\"] = alt_intron_df[\"reference_exon_coord_start\"].astype(int)\n",
    "alt_intron_df[\"reference_exon_coord_end\"] = alt_intron_df[\"reference_exon_coord_end\"].astype(int)\n",
    "alt_intron_df[\"spliced_with_coord_start\"] = alt_intron_df[\"spliced_with_coord_start\"].astype(int)\n",
    "alt_intron_df[\"spliced_with_coord_end\"] = alt_intron_df[\"spliced_with_coord_end\"].astype(int)\n",
    "alt_intron_df[\"junction_coord_start\"] = alt_intron_df[\"junction_coord_start\"].astype(int)\n",
    "alt_intron_df[\"junction_coord_end\"] = alt_intron_df[\"junction_coord_end\"].astype(int)\n",
    "\n",
    "print(alt_intron_df[\"junction_name\"].head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Linhas que violam a teoria:\n",
      "True     27\n",
      "False     4\n",
      "Name: complex, dtype: int64\n",
      "False    104\n",
      "True      81\n",
      "Name: complex, dtype: int64\n",
      "C2    154\n",
      "C1     62\n",
      "Name: spliced_with, dtype: int64 True     185\n",
      "False     31\n",
      "Name: condicao_valida, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "condicao_e2_positivo = (alt_intron_df[\"spliced_with\"] == \"C2\") & (alt_intron_df[\"strand\"] == \"+\") & (alt_intron_df[\"reference_exon_coord_start\"] < alt_intron_df[\"spliced_with_coord_start\"])\n",
    "condicao_e2_negativo = (alt_intron_df[\"spliced_with\"] == \"C2\") & (alt_intron_df[\"strand\"] == \"-\") & (alt_intron_df[\"reference_exon_coord_start\"] > alt_intron_df[\"spliced_with_coord_start\"])\n",
    "condicao_e1_positivo = (alt_intron_df[\"spliced_with\"] == \"C1\") & (alt_intron_df[\"strand\"] == \"+\") & (alt_intron_df[\"reference_exon_coord_start\"] > alt_intron_df[\"spliced_with_coord_start\"])\n",
    "condicao_e1_negativo = (alt_intron_df[\"spliced_with\"] == \"C1\") & (alt_intron_df[\"strand\"] == \"-\") & (alt_intron_df[\"reference_exon_coord_start\"] < alt_intron_df[\"spliced_with_coord_start\"])\n",
    "\n",
    "alt_intron_df[\"condicao_valida\"] = condicao_e2_positivo | condicao_e2_negativo | condicao_e1_positivo | condicao_e1_negativo\n",
    "linhas_invalidas = alt_intron_df[~alt_intron_df[\"condicao_valida\"]]\n",
    "validas = alt_intron_df[alt_intron_df[\"condicao_valida\"]]\n",
    "print(\"Linhas que violam a teoria:\")\n",
    "print(linhas_invalidas[\"complex\"].value_counts())\n",
    "print(validas[\"complex\"].value_counts())\n",
    "print(alt_intron_df[\"spliced_with\"].value_counts(), alt_intron_df[\"condicao_valida\"].value_counts())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Erro de compatibilidade no grupo 2, linha 3:\n",
      "As coordenadas de junção não são compatíveis com reference_exon_coord ou spliced_with_coord.\n",
      "module_id                                       AT4G25080.Araport11.447_1\n",
      "gene_id                                           AT4G25080.Araport11.447\n",
      "gene_name                                                       AT4G25080\n",
      "seqid                                                                Chr4\n",
      "strand                                                                  +\n",
      "lsv_id                        AT4G25080.Araport11.447:t:12878382-12878703\n",
      "event_id                                   AT4G25080.Araport11.447_1_AI_2\n",
      "complex                                                              True\n",
      "denovo                                                              False\n",
      "reference_exon_coord                                    12878339-12878359\n",
      "spliced_with                                                           C1\n",
      "spliced_with_coord                                      12878339-12878359\n",
      "junction_name                                               C2_C1_spliced\n",
      "junction_coord                                          12878145-12878385\n",
      "event_size                                                             22\n",
      "event_non_changing                                                    NaN\n",
      "event_changing                                                        NaN\n",
      "junction_changing                                                     NaN\n",
      "SRR11684912_median_psi                                            0.06757\n",
      "SRR11684912_var_psi                                              0.000151\n",
      "reference_exon_coord_start                                       12878339\n",
      "reference_exon_coord_end                                         12878359\n",
      "spliced_with_coord_start                                         12878339\n",
      "spliced_with_coord_end                                           12878359\n",
      "junction_coord_start                                             12878145\n",
      "junction_coord_end                                               12878385\n",
      "condicao_valida                                                     False\n",
      "Name: 35, dtype: object\n",
      "Erro de compatibilidade no grupo 3, linha 5:\n",
      "As coordenadas de junção não são compatíveis com reference_exon_coord ou spliced_with_coord.\n",
      "module_id                                     AT5G23450.Araport11.447_1\n",
      "gene_id                                         AT5G23450.Araport11.447\n",
      "gene_name                                                     AT5G23450\n",
      "seqid                                                              Chr5\n",
      "strand                                                                -\n",
      "lsv_id                        AT5G23450.Araport11.447:t:7908508-7909050\n",
      "event_id                                 AT5G23450.Araport11.447_1_AI_1\n",
      "complex                                                           False\n",
      "denovo                                                            False\n",
      "reference_exon_coord                                    7909149-7909184\n",
      "spliced_with                                                         C1\n",
      "spliced_with_coord                                      7909149-7909184\n",
      "junction_name                                             C2_C1_spliced\n",
      "junction_coord                                          7909050-7909354\n",
      "event_size                                                           98\n",
      "event_non_changing                                                  NaN\n",
      "event_changing                                                      NaN\n",
      "junction_changing                                                   NaN\n",
      "SRR11684912_median_psi                                           0.2198\n",
      "SRR11684912_var_psi                                            0.002839\n",
      "reference_exon_coord_start                                      7909149\n",
      "reference_exon_coord_end                                        7909184\n",
      "spliced_with_coord_start                                        7909149\n",
      "spliced_with_coord_end                                          7909184\n",
      "junction_coord_start                                            7909050\n",
      "junction_coord_end                                              7909354\n",
      "condicao_valida                                                   False\n",
      "Name: 49, dtype: object\n",
      "Erro de compatibilidade no grupo 4, linha 7:\n",
      "As coordenadas de junção não são compatíveis com reference_exon_coord ou spliced_with_coord.\n",
      "module_id                                       AT1G74960.Araport11.447_1\n",
      "gene_id                                           AT1G74960.Araport11.447\n",
      "gene_name                                                       AT1G74960\n",
      "seqid                                                                Chr1\n",
      "strand                                                                  -\n",
      "lsv_id                        AT1G74960.Araport11.447:t:28151920-28152241\n",
      "event_id                                   AT1G74960.Araport11.447_1_AI_1\n",
      "complex                                                             False\n",
      "denovo                                                              False\n",
      "reference_exon_coord                                    28152253-28152273\n",
      "spliced_with                                                           C1\n",
      "spliced_with_coord                                      28152253-28152273\n",
      "junction_name                                               C2_C1_spliced\n",
      "junction_coord                                          28152241-28152512\n",
      "event_size                                                             11\n",
      "event_non_changing                                                    NaN\n",
      "event_changing                                                        NaN\n",
      "junction_changing                                                     NaN\n",
      "SRR11684912_median_psi                                            0.05172\n",
      "SRR11684912_var_psi                                              0.000334\n",
      "reference_exon_coord_start                                       28152253\n",
      "reference_exon_coord_end                                         28152273\n",
      "spliced_with_coord_start                                         28152253\n",
      "spliced_with_coord_end                                           28152273\n",
      "junction_coord_start                                             28152241\n",
      "junction_coord_end                                               28152512\n",
      "condicao_valida                                                     False\n",
      "Name: 71, dtype: object\n",
      "Erro de compatibilidade no grupo 5, linha 9:\n",
      "As coordenadas de junção não são compatíveis com reference_exon_coord ou spliced_with_coord.\n",
      "module_id                                       AT3G28550.Araport11.447_1\n",
      "gene_id                                           AT3G28550.Araport11.447\n",
      "gene_name                                                       AT3G28550\n",
      "seqid                                                                Chr3\n",
      "strand                                                                  -\n",
      "lsv_id                        AT3G28550.Araport11.447:t:10700562-10701607\n",
      "event_id                                   AT3G28550.Araport11.447_1_AI_2\n",
      "complex                                                              True\n",
      "denovo                                                               True\n",
      "reference_exon_coord                                    10701608-10701682\n",
      "spliced_with                                                           C1\n",
      "spliced_with_coord                                      10701608-10701682\n",
      "junction_name                                               C2_C1_spliced\n",
      "junction_coord                                          10701607-10701683\n",
      "event_size                                                              2\n",
      "event_non_changing                                                    NaN\n",
      "event_changing                                                        NaN\n",
      "junction_changing                                                     NaN\n",
      "SRR11684912_median_psi                                             0.3775\n",
      "SRR11684912_var_psi                                              0.005886\n",
      "reference_exon_coord_start                                       10701608\n",
      "reference_exon_coord_end                                         10701682\n",
      "spliced_with_coord_start                                         10701608\n",
      "spliced_with_coord_end                                           10701682\n",
      "junction_coord_start                                             10701607\n",
      "junction_coord_end                                               10701683\n",
      "condicao_valida                                                     False\n",
      "Name: 75, dtype: object\n",
      "Erro de compatibilidade no grupo 6, linha 11:\n",
      "As coordenadas de junção não são compatíveis com reference_exon_coord ou spliced_with_coord.\n",
      "module_id                                       AT3G28550.Araport11.447_1\n",
      "gene_id                                           AT3G28550.Araport11.447\n",
      "gene_name                                                       AT3G28550\n",
      "seqid                                                                Chr3\n",
      "strand                                                                  -\n",
      "lsv_id                        AT3G28550.Araport11.447:t:10701683-10701757\n",
      "event_id                                   AT3G28550.Araport11.447_1_AI_5\n",
      "complex                                                              True\n",
      "denovo                                                               True\n",
      "reference_exon_coord                                    10701758-10701858\n",
      "spliced_with                                                           C1\n",
      "spliced_with_coord                                      10701758-10701858\n",
      "junction_name                                               C2_C1_spliced\n",
      "junction_coord                                          10701708-10702285\n",
      "event_size                                                              2\n",
      "event_non_changing                                                    NaN\n",
      "event_changing                                                        NaN\n",
      "junction_changing                                                     NaN\n",
      "SRR11684912_median_psi                                             0.1655\n",
      "SRR11684912_var_psi                                              0.003562\n",
      "reference_exon_coord_start                                       10701758\n",
      "reference_exon_coord_end                                         10701858\n",
      "spliced_with_coord_start                                         10701758\n",
      "spliced_with_coord_end                                           10701858\n",
      "junction_coord_start                                             10701708\n",
      "junction_coord_end                                               10702285\n",
      "condicao_valida                                                     False\n",
      "Name: 81, dtype: object\n",
      "Erro de compatibilidade no grupo 8, linha 15:\n",
      "As coordenadas de junção não são compatíveis com reference_exon_coord ou spliced_with_coord.\n",
      "module_id                                       AT3G28550.Araport11.447_1\n",
      "gene_id                                           AT3G28550.Araport11.447\n",
      "gene_name                                                       AT3G28550\n",
      "seqid                                                                Chr3\n",
      "strand                                                                  -\n",
      "lsv_id                        AT3G28550.Araport11.447:t:10701908-10702183\n",
      "event_id                                  AT3G28550.Araport11.447_1_AI_10\n",
      "complex                                                              True\n",
      "denovo                                                               True\n",
      "reference_exon_coord                                    10702184-10702258\n",
      "spliced_with                                                           C1\n",
      "spliced_with_coord                                      10702184-10702258\n",
      "junction_name                                               C2_C1_spliced\n",
      "junction_coord                                          10702183-10702259\n",
      "event_size                                                              2\n",
      "event_non_changing                                                    NaN\n",
      "event_changing                                                        NaN\n",
      "junction_changing                                                     NaN\n",
      "SRR11684912_median_psi                                             0.2348\n",
      "SRR11684912_var_psi                                              0.006174\n",
      "reference_exon_coord_start                                       10702184\n",
      "reference_exon_coord_end                                         10702258\n",
      "spliced_with_coord_start                                         10702184\n",
      "spliced_with_coord_end                                           10702258\n",
      "junction_coord_start                                             10702183\n",
      "junction_coord_end                                               10702259\n",
      "condicao_valida                                                     False\n",
      "Name: 91, dtype: object\n",
      "Erro de compatibilidade no grupo 9, linha 17:\n",
      "As coordenadas de junção não são compatíveis com reference_exon_coord ou spliced_with_coord.\n",
      "module_id                                     AT4G05320.Araport11.447_1\n",
      "gene_id                                         AT4G05320.Araport11.447\n",
      "gene_name                                                     AT4G05320\n",
      "seqid                                                              Chr4\n",
      "strand                                                                +\n",
      "lsv_id                        AT4G05320.Araport11.447:t:2719469-2719551\n",
      "event_id                                 AT4G05320.Araport11.447_1_AI_6\n",
      "complex                                                            True\n",
      "denovo                                                             True\n",
      "reference_exon_coord                                    2719291-2719468\n",
      "spliced_with                                                         C1\n",
      "spliced_with_coord                                      2719291-2719468\n",
      "junction_name                                             C2_C1_spliced\n",
      "junction_coord                                          2719240-2719469\n",
      "event_size                                                            2\n",
      "event_non_changing                                                  NaN\n",
      "event_changing                                                      NaN\n",
      "junction_changing                                                   NaN\n",
      "SRR11684912_median_psi                                           0.1897\n",
      "SRR11684912_var_psi                                            0.000313\n",
      "reference_exon_coord_start                                      2719291\n",
      "reference_exon_coord_end                                        2719468\n",
      "spliced_with_coord_start                                        2719291\n",
      "spliced_with_coord_end                                          2719468\n",
      "junction_coord_start                                            2719240\n",
      "junction_coord_end                                              2719469\n",
      "condicao_valida                                                   False\n",
      "Name: 127, dtype: object\n",
      "Erro de compatibilidade no grupo 10, linha 19:\n",
      "As coordenadas de junção não são compatíveis com reference_exon_coord ou spliced_with_coord.\n",
      "module_id                                     AT4G05320.Araport11.447_1\n",
      "gene_id                                         AT4G05320.Araport11.447\n",
      "gene_name                                                     AT4G05320\n",
      "seqid                                                              Chr4\n",
      "strand                                                                +\n",
      "lsv_id                        AT4G05320.Araport11.447:t:2719552-2720308\n",
      "event_id                                 AT4G05320.Araport11.447_1_AI_8\n",
      "complex                                                            True\n",
      "denovo                                                            False\n",
      "reference_exon_coord                                    2719469-2719551\n",
      "spliced_with                                                         C1\n",
      "spliced_with_coord                                      2719469-2719551\n",
      "junction_name                                             C2_C1_spliced\n",
      "junction_coord                                          2719468-2719697\n",
      "event_size                                                            2\n",
      "event_non_changing                                                  NaN\n",
      "event_changing                                                      NaN\n",
      "junction_changing                                                   NaN\n",
      "SRR11684912_median_psi                                          0.05258\n",
      "SRR11684912_var_psi                                            0.000137\n",
      "reference_exon_coord_start                                      2719469\n",
      "reference_exon_coord_end                                        2719551\n",
      "spliced_with_coord_start                                        2719469\n",
      "spliced_with_coord_end                                          2719551\n",
      "junction_coord_start                                            2719468\n",
      "junction_coord_end                                              2719697\n",
      "condicao_valida                                                   False\n",
      "Name: 131, dtype: object\n",
      "Erro de compatibilidade no grupo 11, linha 21:\n",
      "As coordenadas de junção não são compatíveis com reference_exon_coord ou spliced_with_coord.\n",
      "module_id                                       AT5G39570.Araport11.447_1\n",
      "gene_id                                           AT5G39570.Araport11.447\n",
      "gene_name                                                       AT5G39570\n",
      "seqid                                                                Chr5\n",
      "strand                                                                  +\n",
      "lsv_id                        AT5G39570.Araport11.447:t:15844749-15845076\n",
      "event_id                                   AT5G39570.Araport11.447_1_AI_3\n",
      "complex                                                              True\n",
      "denovo                                                              False\n",
      "reference_exon_coord                                    15844679-15844735\n",
      "spliced_with                                                           C1\n",
      "spliced_with_coord                                      15844679-15844735\n",
      "junction_name                                               C2_C1_spliced\n",
      "junction_coord                                          15844633-15844772\n",
      "event_size                                                             13\n",
      "event_non_changing                                                    NaN\n",
      "event_changing                                                        NaN\n",
      "junction_changing                                                     NaN\n",
      "SRR11684912_median_psi                                            0.06719\n",
      "SRR11684912_var_psi                                              0.000156\n",
      "reference_exon_coord_start                                       15844679\n",
      "reference_exon_coord_end                                         15844735\n",
      "spliced_with_coord_start                                         15844679\n",
      "spliced_with_coord_end                                           15844735\n",
      "junction_coord_start                                             15844633\n",
      "junction_coord_end                                               15844772\n",
      "condicao_valida                                                     False\n",
      "Name: 145, dtype: object\n",
      "Erro de compatibilidade no grupo 12, linha 23:\n",
      "As coordenadas de junção não são compatíveis com reference_exon_coord ou spliced_with_coord.\n",
      "module_id                                     AT1G15830.Araport11.447_1\n",
      "gene_id                                         AT1G15830.Araport11.447\n",
      "gene_name                                                     AT1G15830\n",
      "seqid                                                              Chr1\n",
      "strand                                                                +\n",
      "lsv_id                        AT1G15830.Araport11.447:t:5449213-5449308\n",
      "event_id                                 AT1G15830.Araport11.447_1_AI_6\n",
      "complex                                                            True\n",
      "denovo                                                             True\n",
      "reference_exon_coord                                    5449204-5449212\n",
      "spliced_with                                                         C1\n",
      "spliced_with_coord                                      5449204-5449212\n",
      "junction_name                                             C2_C1_spliced\n",
      "junction_coord                                          5449164-5449213\n",
      "event_size                                                            2\n",
      "event_non_changing                                                  NaN\n",
      "event_changing                                                      NaN\n",
      "junction_changing                                                   NaN\n",
      "SRR11684912_median_psi                                           0.2981\n",
      "SRR11684912_var_psi                                             0.00301\n",
      "reference_exon_coord_start                                      5449204\n",
      "reference_exon_coord_end                                        5449212\n",
      "spliced_with_coord_start                                        5449204\n",
      "spliced_with_coord_end                                          5449212\n",
      "junction_coord_start                                            5449164\n",
      "junction_coord_end                                              5449213\n",
      "condicao_valida                                                   False\n",
      "Name: 173, dtype: object\n",
      "Erro de compatibilidade no grupo 14, linha 27:\n",
      "As coordenadas de junção não são compatíveis com reference_exon_coord ou spliced_with_coord.\n",
      "module_id                                     AT1G07930.Araport11.447_1\n",
      "gene_id                                         AT1G07930.Araport11.447\n",
      "gene_name                                                     AT1G07930\n",
      "seqid                                                              Chr1\n",
      "strand                                                                +\n",
      "lsv_id                        AT1G07930.Araport11.447:t:2460102-2460817\n",
      "event_id                                 AT1G07930.Araport11.447_1_AI_1\n",
      "complex                                                            True\n",
      "denovo                                                             True\n",
      "reference_exon_coord                                    2459571-2459870\n",
      "spliced_with                                                         C1\n",
      "spliced_with_coord                                      2459571-2459870\n",
      "junction_name                                             C2_C1_spliced\n",
      "junction_coord                                          2456709-2460167\n",
      "event_size                                                          231\n",
      "event_non_changing                                                  NaN\n",
      "event_changing                                                      NaN\n",
      "junction_changing                                                   NaN\n",
      "SRR11684912_median_psi                                           0.3242\n",
      "SRR11684912_var_psi                                            0.000411\n",
      "reference_exon_coord_start                                      2459571\n",
      "reference_exon_coord_end                                        2459870\n",
      "spliced_with_coord_start                                        2459571\n",
      "spliced_with_coord_end                                          2459870\n",
      "junction_coord_start                                            2456709\n",
      "junction_coord_end                                              2460167\n",
      "condicao_valida                                                   False\n",
      "Name: 203, dtype: object\n",
      "Erro de compatibilidade no grupo 15, linha 29:\n",
      "As coordenadas de junção não são compatíveis com reference_exon_coord ou spliced_with_coord.\n",
      "module_id                                       AT2G25970.Araport11.447_1\n",
      "gene_id                                           AT2G25970.Araport11.447\n",
      "gene_name                                                       AT2G25970\n",
      "seqid                                                                Chr2\n",
      "strand                                                                  -\n",
      "lsv_id                        AT2G25970.Araport11.447:t:11071622-11072830\n",
      "event_id                                   AT2G25970.Araport11.447_1_AI_1\n",
      "complex                                                              True\n",
      "denovo                                                               True\n",
      "reference_exon_coord                                    11074353-11074472\n",
      "spliced_with                                                           C1\n",
      "spliced_with_coord                                      11074353-11074472\n",
      "junction_name                                               C2_C1_spliced\n",
      "junction_coord                                          11072830-11073061\n",
      "event_size                                                           1522\n",
      "event_non_changing                                                    NaN\n",
      "event_changing                                                        NaN\n",
      "junction_changing                                                     NaN\n",
      "SRR11684912_median_psi                                            0.05534\n",
      "SRR11684912_var_psi                                              0.000686\n",
      "reference_exon_coord_start                                       11074353\n",
      "reference_exon_coord_end                                         11074472\n",
      "spliced_with_coord_start                                         11074353\n",
      "spliced_with_coord_end                                           11074472\n",
      "junction_coord_start                                             11072830\n",
      "junction_coord_end                                               11073061\n",
      "condicao_valida                                                     False\n",
      "Name: 207, dtype: object\n"
     ]
    },
    {
     "ename": "IndexError",
     "evalue": "single positional indexer is out-of-bounds",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[69], line 22\u001b[0m\n\u001b[1;32m     19\u001b[0m                 \u001b[38;5;28mprint\u001b[39m(splice_row)\n\u001b[1;32m     21\u001b[0m \u001b[38;5;66;03m# Aplicar a função ao DataFrame\u001b[39;00m\n\u001b[0;32m---> 22\u001b[0m \u001b[43mtest_conditions\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlinhas_invalidas\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[69], line 7\u001b[0m, in \u001b[0;36mtest_conditions\u001b[0;34m(df)\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;66;03m# 2) Verifica compatibilidade das coordenadas nas linhas marcadas como splice\u001b[39;00m\n\u001b[1;32m      6\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m j \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;241m1\u001b[39m, \u001b[38;5;241m2\u001b[39m, \u001b[38;5;241m2\u001b[39m):  \u001b[38;5;66;03m# Apenas linhas ímpares (splice)\u001b[39;00m\n\u001b[0;32m----> 7\u001b[0m     splice_row \u001b[38;5;241m=\u001b[39m \u001b[43mgroup\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43miloc\u001b[49m\u001b[43m[\u001b[49m\u001b[43mj\u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m      9\u001b[0m     \u001b[38;5;66;03m# Extrai coordenadas\u001b[39;00m\n\u001b[1;32m     10\u001b[0m     junction_start, junction_end \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mmap\u001b[39m(\u001b[38;5;28mint\u001b[39m, splice_row[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mjunction_coord\u001b[39m\u001b[38;5;124m\"\u001b[39m]\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m-\u001b[39m\u001b[38;5;124m\"\u001b[39m))\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/pandas/core/indexing.py:967\u001b[0m, in \u001b[0;36m_LocationIndexer.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    964\u001b[0m axis \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maxis \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;241m0\u001b[39m\n\u001b[1;32m    966\u001b[0m maybe_callable \u001b[38;5;241m=\u001b[39m com\u001b[38;5;241m.\u001b[39mapply_if_callable(key, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj)\n\u001b[0;32m--> 967\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_getitem_axis\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmaybe_callable\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/pandas/core/indexing.py:1520\u001b[0m, in \u001b[0;36m_iLocIndexer._getitem_axis\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1517\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCannot index by location index with a non-integer key\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   1519\u001b[0m \u001b[38;5;66;03m# validate the location\u001b[39;00m\n\u001b[0;32m-> 1520\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_validate_integer\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1522\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj\u001b[38;5;241m.\u001b[39m_ixs(key, axis\u001b[38;5;241m=\u001b[39maxis)\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/pandas/core/indexing.py:1452\u001b[0m, in \u001b[0;36m_iLocIndexer._validate_integer\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1450\u001b[0m len_axis \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlen\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj\u001b[38;5;241m.\u001b[39m_get_axis(axis))\n\u001b[1;32m   1451\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m key \u001b[38;5;241m>\u001b[39m\u001b[38;5;241m=\u001b[39m len_axis \u001b[38;5;129;01mor\u001b[39;00m key \u001b[38;5;241m<\u001b[39m \u001b[38;5;241m-\u001b[39mlen_axis:\n\u001b[0;32m-> 1452\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mIndexError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msingle positional indexer is out-of-bounds\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mIndexError\u001b[0m: single positional indexer is out-of-bounds"
     ]
    }
   ],
   "source": [
    "def test_conditions(df):\n",
    "    for i in range(0, len(df), 2):  # Processa grupos de 4 linhas\n",
    "        group = df.iloc[i:i+2]  # Grupo de 4 linhas\n",
    "\n",
    "        # 2) Verifica compatibilidade das coordenadas nas linhas marcadas como splice\n",
    "        for j in range(1, 2, 2):  # Apenas linhas ímpares (splice)\n",
    "            splice_row = group.iloc[j]\n",
    "\n",
    "            # Extrai coordenadas\n",
    "            junction_start, junction_end = map(int, splice_row[\"junction_coord\"].split(\"-\"))\n",
    "            ref_start, ref_end = map(int, splice_row[\"reference_exon_coord\"].split(\"-\"))\n",
    "            spliced_start, spliced_end = map(int, splice_row[\"spliced_with_coord\"].split(\"-\"))\n",
    "\n",
    "            # Verifica compatibilidade\n",
    "            if not (junction_start in [ref_start, ref_end, spliced_start, spliced_end] or\n",
    "                    junction_end in [ref_start, ref_end, spliced_start, spliced_end]):\n",
    "                print(f\"Erro de compatibilidade no grupo {i//2 + 1}, linha {i + j}:\")\n",
    "                print(\"As coordenadas de junção não são compatíveis com reference_exon_coord ou spliced_with_coord.\")\n",
    "                print(splice_row)\n",
    "\n",
    "# Aplicar a função ao DataFrame\n",
    "test_conditions(linhas_invalidas)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cassette or Exon Skipping\n",
    "Exon skipping, also known as cassette exon, is a form of alternative splicing where an exon is either included or excluded from the final mRNA transcript. This process allows a single gene to produce multiple protein isoforms, contributing to the diversity of the proteome.\n",
    "\n",
    "- Organized in stacks of 4 rows, where junction_names are respectively: C1_C2, C1_A, C2_C1, C2_A. \n",
    "\n",
    "- !! TODO: C1_C2 and C2_C1 are the same, but represent different LSVs, which means they have different mean PSI values. \n",
    "\n",
    "- For + strand: \n",
    "    - C1_C2: reference_exon_coord_start < spliced_with_coord_start\n",
    "    - C1_A: reference_exon_coord_start < spliced_with_coord_start\n",
    "    - C2_C1: reference_exon_coord_start > spliced_with_coord_start\n",
    "    - C2_A: reference_exon_coord_start > spliced_with_coord_start\n",
    "\n",
    "- For - strand:\n",
    "    - C1_C2: reference_exon_coord_start > spliced_with_coord_start\n",
    "    - C1_A: reference_exon_coord_start > spliced_with_coord_start\n",
    "    - C2_C1: reference_exon_coord_start < spliced_with_coord_start\n",
    "    - C2_A: reference_exon_coord_start < spliced_with_coord_start"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                   module_id                  gene_id  gene_name seqid strand  \\\n",
      "0  AT1G06220.Araport11.447_1  AT1G06220.Araport11.447  AT1G06220  Chr1      +   \n",
      "1  AT1G06220.Araport11.447_1  AT1G06220.Araport11.447  AT1G06220  Chr1      +   \n",
      "2  AT1G06220.Araport11.447_1  AT1G06220.Araport11.447  AT1G06220  Chr1      +   \n",
      "3  AT1G06220.Araport11.447_1  AT1G06220.Araport11.447  AT1G06220  Chr1      +   \n",
      "4  AT2G26770.Araport11.447_1  AT2G26770.Araport11.447  AT2G26770  Chr2      -   \n",
      "\n",
      "                                        lsv_id  \\\n",
      "0    AT1G06220.Araport11.447:s:1899951-1900107   \n",
      "1    AT1G06220.Araport11.447:s:1899951-1900107   \n",
      "2    AT1G06220.Araport11.447:t:1900512-1901264   \n",
      "3    AT1G06220.Araport11.447:t:1900512-1901264   \n",
      "4  AT2G26770.Araport11.447:s:11408200-11408421   \n",
      "\n",
      "                         event_id  complex  denovo reference_exon_coord  ...  \\\n",
      "0  AT1G06220.Araport11.447_1_CE_1    False   False      1899951-1900107  ...   \n",
      "1  AT1G06220.Araport11.447_1_CE_1    False   False      1899951-1900107  ...   \n",
      "2  AT1G06220.Araport11.447_1_CE_1    False   False      1900512-1901264  ...   \n",
      "3  AT1G06220.Araport11.447_1_CE_1    False   False      1900512-1901264  ...   \n",
      "4  AT2G26770.Araport11.447_1_CE_1     True   False    11408200-11408421  ...   \n",
      "\n",
      "  event_changing junction_changing SRR11684912_median_psi SRR11684912_var_psi  \\\n",
      "0            NaN               NaN                 0.8523            0.002481   \n",
      "1            NaN               NaN                 0.1477            0.002481   \n",
      "2            NaN               NaN                 0.8630            0.001700   \n",
      "3            NaN               NaN                 0.1370            0.001700   \n",
      "4            NaN               NaN                 0.6296            0.005735   \n",
      "\n",
      "   reference_exon_coord_start reference_exon_coord_end  \\\n",
      "0                     1899951                  1900107   \n",
      "1                     1899951                  1900107   \n",
      "2                     1900512                  1901264   \n",
      "3                     1900512                  1901264   \n",
      "4                    11408200                 11408421   \n",
      "\n",
      "  spliced_with_coord_start  spliced_with_coord_end  junction_coord_start  \\\n",
      "0                  1900512                 1901264               1900107   \n",
      "1                  1900297                 1900429               1900107   \n",
      "2                  1899951                 1900107               1900107   \n",
      "3                  1900297                 1900429               1900429   \n",
      "4                 11407462                11407748              11407748   \n",
      "\n",
      "   junction_coord_end  \n",
      "0             1900512  \n",
      "1             1900297  \n",
      "2             1900512  \n",
      "3             1900512  \n",
      "4            11408200  \n",
      "\n",
      "[5 rows x 26 columns]\n"
     ]
    }
   ],
   "source": [
    "cassette_df = pd.read_csv(cassette, sep=\"\\t\", header=0, comment=\"#\")\n",
    "\n",
    "cassette_df[[\"reference_exon_coord_start\", \"reference_exon_coord_end\"]] = cassette_df[\"reference_exon_coord\"].str.split(\"-\", expand=True)\n",
    "cassette_df[[\"spliced_with_coord_start\", \"spliced_with_coord_end\"]] = cassette_df[\"spliced_with_coord\"].str.split(\"-\", expand=True)\n",
    "cassette_df[[\"junction_coord_start\", \"junction_coord_end\"]] = cassette_df[\"junction_coord\"].str.split(\"-\", expand=True)\n",
    "\n",
    "# convert start and end to int\n",
    "cassette_df[\"reference_exon_coord_start\"] = cassette_df[\"reference_exon_coord_start\"].astype(int)\n",
    "cassette_df[\"reference_exon_coord_end\"] = cassette_df[\"reference_exon_coord_end\"].astype(int)\n",
    "cassette_df[\"spliced_with_coord_start\"] = cassette_df[\"spliced_with_coord_start\"].astype(int)\n",
    "cassette_df[\"spliced_with_coord_end\"] = cassette_df[\"spliced_with_coord_end\"].astype(int)\n",
    "cassette_df[\"junction_coord_start\"] = cassette_df[\"junction_coord_start\"].astype(int)\n",
    "cassette_df[\"junction_coord_end\"] = cassette_df[\"junction_coord_end\"].astype(int)\n",
    "\n",
    "print(cassette_df.head())\n",
    "\n",
    "def check_junction_name_order(df):\n",
    "    expected_order = [\"C1_C2\", \"C1_A\", \"C2_C1\", \"C2_A\"]\n",
    "    \n",
    "    for i in range(0, len(df), 4):  # Processa grupos de 4 linhas\n",
    "        group = df.iloc[i:i+4]  # Grupo de 4 linhas\n",
    "        junction_names = group[\"junction_name\"].tolist()  # Lista de valores na coluna junction_name\n",
    "\n",
    "        if junction_names != expected_order:\n",
    "            print(f\"Erro de ordem no grupo {i//4 + 1}:\")\n",
    "            print(f\"Esperado: {expected_order}\")\n",
    "            print(f\"Encontrado: {junction_names}\")\n",
    "            print(group)\n",
    "            continue  # Pula para o próximo grupo\n",
    "\n",
    "# Aplicar a função ao DataFrame\n",
    "check_junction_name_order(cassette_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Linhas que violam a teoria:\n",
      "Empty DataFrame\n",
      "Columns: [module_id, gene_id, gene_name, seqid, strand, lsv_id, event_id, complex, denovo, reference_exon_coord, spliced_with, spliced_with_coord, junction_name, junction_coord, event_size, event_non_changing, event_changing, junction_changing, SRR11684912_median_psi, SRR11684912_var_psi, reference_exon_coord_start, reference_exon_coord_end, spliced_with_coord_start, spliced_with_coord_end, junction_coord_start, junction_coord_end, condicao_valida]\n",
      "Index: []\n",
      "\n",
      "[0 rows x 27 columns]\n",
      "\n",
      "Contagem de valores em 'condicao_valida':\n",
      "True    884\n",
      "Name: condicao_valida, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "def check_coordinate_conditions(df):\n",
    "    # Condições para strand +\n",
    "    condicao_c1_c2_positivo = (df[\"junction_name\"].isin([\"C1_C2\", \"C1_A\"])) & (df[\"strand\"] == \"+\") & (df[\"reference_exon_coord_start\"] < df[\"spliced_with_coord_start\"])\n",
    "    condicao_c2_c1_positivo = (df[\"junction_name\"].isin([\"C2_C1\", \"C2_A\"])) & (df[\"strand\"] == \"+\") & (df[\"reference_exon_coord_start\"] > df[\"spliced_with_coord_start\"])\n",
    "\n",
    "    # Condições para strand -\n",
    "    condicao_c1_c2_negativo = (df[\"junction_name\"].isin([\"C1_C2\", \"C1_A\"])) & (df[\"strand\"] == \"-\") & (df[\"reference_exon_coord_start\"] > df[\"spliced_with_coord_start\"])\n",
    "    condicao_c2_c1_negativo = (df[\"junction_name\"].isin([\"C2_C1\", \"C2_A\"])) & (df[\"strand\"] == \"-\") & (df[\"reference_exon_coord_start\"] < df[\"spliced_with_coord_start\"])\n",
    "\n",
    "    # Combina todas as condições\n",
    "    df[\"condicao_valida\"] = condicao_c1_c2_positivo | condicao_c2_c1_positivo | condicao_c1_c2_negativo | condicao_c2_c1_negativo\n",
    "\n",
    "    # Identifica linhas inválidas\n",
    "    linhas_invalidas = df[~df[\"condicao_valida\"]]\n",
    "    print(\"Linhas que violam a teoria:\")\n",
    "    print(linhas_invalidas)\n",
    "\n",
    "    # Estatísticas\n",
    "    print(\"\\nContagem de valores em 'condicao_valida':\")\n",
    "    print(df[\"condicao_valida\"].value_counts())\n",
    "\n",
    "# Aplicar a função ao DataFrame\n",
    "check_coordinate_conditions(cassette_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tandem Cassette Exons\n",
    "Same as Cassette or Exon Skipping, but with more than 2 exons being skipped. \n",
    "\n",
    "- Organized in stacks of 4 rows, where junction_names are respectively: C1_C2, C1_A, C2_C1, C2_A_Last. \n",
    "\n",
    "- !! TODO: C1_C2 and C2_C1 are the same, but represent different LSVs, which means they have different mean PSI values. \n",
    "\n",
    "- For + strand: \n",
    "    - C1_C2: reference_exon_coord_start < spliced_with_coord_start\n",
    "    - C1_A: reference_exon_coord_start < spliced_with_coord_start\n",
    "    - C2_C1: reference_exon_coord_start > spliced_with_coord_start\n",
    "    - C2_A_Last: reference_exon_coord_start > spliced_with_coord_start\n",
    "\n",
    "- For - strand:\n",
    "    - C1_C2: reference_exon_coord_start > spliced_with_coord_start\n",
    "    - C1_A: reference_exon_coord_start > spliced_with_coord_start\n",
    "    - C2_C1: reference_exon_coord_start < spliced_with_coord_start\n",
    "    - C2_A_Last: reference_exon_coord_start < spliced_with_coord_start\n",
    "\n",
    "- !! TODO: Has two new columns (exons_skipped_coords, num_skipped_exons) that actually disrup the pattern of the columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                   module_id                  gene_id  gene_name seqid strand  \\\n",
      "0  AT2G39950.Araport11.447_1  AT2G39950.Araport11.447  AT2G39950  Chr2      -   \n",
      "1  AT2G39950.Araport11.447_1  AT2G39950.Araport11.447  AT2G39950  Chr2      -   \n",
      "2  AT2G39950.Araport11.447_1  AT2G39950.Araport11.447  AT2G39950  Chr2      -   \n",
      "3  AT2G39950.Araport11.447_1  AT2G39950.Araport11.447  AT2G39950  Chr2      -   \n",
      "4  AT1G60640.Araport11.447_1  AT1G60640.Araport11.447  AT1G60640  Chr1      -   \n",
      "\n",
      "                                        lsv_id  \\\n",
      "0  AT2G39950.Araport11.447:s:16679657-16679772   \n",
      "1  AT2G39950.Araport11.447:s:16679657-16679772   \n",
      "2  AT2G39950.Araport11.447:t:16678035-16678278   \n",
      "3  AT2G39950.Araport11.447:t:16678035-16678278   \n",
      "4  AT1G60640.Araport11.447:s:22338467-22338526   \n",
      "\n",
      "                          event_id  complex  denovo reference_exon_coord  ...  \\\n",
      "0  AT2G39950.Araport11.447_1_TCE_1     True   False    16679657-16679772  ...   \n",
      "1  AT2G39950.Araport11.447_1_TCE_1     True   False    16679657-16679772  ...   \n",
      "2  AT2G39950.Araport11.447_1_TCE_1     True   False    16678035-16678278  ...   \n",
      "3  AT2G39950.Araport11.447_1_TCE_1     True   False    16678035-16678278  ...   \n",
      "4  AT1G60640.Araport11.447_1_TCE_1     True   False    22338467-22338526  ...   \n",
      "\n",
      "  event_changing junction_changing SRR11684912_median_psi  \\\n",
      "0            NaN               NaN                0.71120   \n",
      "1            NaN               NaN                0.08627   \n",
      "2            NaN               NaN                0.64970   \n",
      "3            NaN               NaN                0.35030   \n",
      "4            NaN               NaN                0.12320   \n",
      "\n",
      "   SRR11684912_var_psi reference_exon_coord_start reference_exon_coord_end  \\\n",
      "0             0.005723                   16679657                 16679772   \n",
      "1             0.002168                   16679657                 16679772   \n",
      "2             0.004900                   16678035                 16678278   \n",
      "3             0.004900                   16678035                 16678278   \n",
      "4             0.007875                   22338467                 22338526   \n",
      "\n",
      "   spliced_with_coord_start  spliced_with_coord_end  junction_coord_start  \\\n",
      "0                  16678035                16678278              16678278   \n",
      "1                  16678967                16679015              16679426   \n",
      "2                  16679657                16679772              16678278   \n",
      "3                  16679173                16679426              16678278   \n",
      "4                  22337701                22337766              22337766   \n",
      "\n",
      "   junction_coord_end  \n",
      "0            16679657  \n",
      "1            16679657  \n",
      "2            16679657  \n",
      "3            16678967  \n",
      "4            22338467  \n",
      "\n",
      "[5 rows x 28 columns]\n"
     ]
    }
   ],
   "source": [
    "tandem_cassette_df = pd.read_csv(tandem_cassette, sep=\"\\t\", header=0, comment=\"#\")\n",
    "\n",
    "tandem_cassette_df[[\"reference_exon_coord_start\", \"reference_exon_coord_end\"]] = tandem_cassette_df[\"reference_exon_coord\"].str.split(\"-\", expand=True)\n",
    "tandem_cassette_df[[\"spliced_with_coord_start\", \"spliced_with_coord_end\"]] = tandem_cassette_df[\"spliced_with_coord\"].str.split(\"-\", expand=True)\n",
    "tandem_cassette_df[[\"junction_coord_start\", \"junction_coord_end\"]] = tandem_cassette_df[\"junction_coord\"].str.split(\"-\", expand=True)\n",
    "\n",
    "# convert start and end to int\n",
    "tandem_cassette_df[\"reference_exon_coord_start\"] = tandem_cassette_df[\"reference_exon_coord_start\"].astype(int)\n",
    "tandem_cassette_df[\"reference_exon_coord_end\"] = tandem_cassette_df[\"reference_exon_coord_end\"].astype(int)\n",
    "tandem_cassette_df[\"spliced_with_coord_start\"] = tandem_cassette_df[\"spliced_with_coord_start\"].astype(int)\n",
    "tandem_cassette_df[\"spliced_with_coord_end\"] = tandem_cassette_df[\"spliced_with_coord_end\"].astype(int)\n",
    "tandem_cassette_df[\"junction_coord_start\"] = tandem_cassette_df[\"junction_coord_start\"].astype(int)\n",
    "tandem_cassette_df[\"junction_coord_end\"] = tandem_cassette_df[\"junction_coord_end\"].astype(int)\n",
    "\n",
    "print(tandem_cassette_df.head())\n",
    "\n",
    "def check_junction_name_order(df):\n",
    "    expected_order = [\"C1_C2\", \"C1_A\", \"C2_C1\", \"C2_A_Last\"]\n",
    "    \n",
    "    for i in range(0, len(df), 4):  # Processa grupos de 4 linhas\n",
    "        group = df.iloc[i:i+4]  # Grupo de 4 linhas\n",
    "        junction_names = group[\"junction_name\"].tolist()  # Lista de valores na coluna junction_name\n",
    "\n",
    "        if junction_names != expected_order:\n",
    "            print(f\"Erro de ordem no grupo {i//4 + 1}:\")\n",
    "            print(f\"Esperado: {expected_order}\")\n",
    "            print(f\"Encontrado: {junction_names}\")\n",
    "            print(group)\n",
    "            continue  # Pula para o próximo grupo\n",
    "\n",
    "# Aplicar a função ao DataFrame\n",
    "check_junction_name_order(tandem_cassette_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Linhas que violam a teoria:\n",
      "Empty DataFrame\n",
      "Columns: [reference_exon_coord, spliced_with_coord, junction_coord, junction_name, strand]\n",
      "Index: []\n",
      "\n",
      "Contagem de valores em 'condicao_valida':\n",
      "True    32\n",
      "Name: condicao_valida, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "def check_coordinate_conditions(df):\n",
    "    # Condições para strand +\n",
    "    condicao_c1_c2_positivo = (df[\"junction_name\"].isin([\"C1_C2\", \"C1_A\"])) & (df[\"strand\"] == \"+\") & (df[\"reference_exon_coord_start\"] < df[\"spliced_with_coord_start\"])\n",
    "    condicao_c2_c1_positivo = (df[\"junction_name\"].isin([\"C2_C1\", \"C2_A_Last\"])) & (df[\"strand\"] == \"+\") & (df[\"reference_exon_coord_start\"] > df[\"spliced_with_coord_start\"])\n",
    "\n",
    "    # Condições para strand -\n",
    "    condicao_c1_c2_negativo = (df[\"junction_name\"].isin([\"C1_C2\", \"C1_A\"])) & (df[\"strand\"] == \"-\") & (df[\"reference_exon_coord_start\"] > df[\"spliced_with_coord_start\"])\n",
    "    condicao_c2_c1_negativo = (df[\"junction_name\"].isin([\"C2_C1\", \"C2_A_Last\"])) & (df[\"strand\"] == \"-\") & (df[\"reference_exon_coord_start\"] < df[\"spliced_with_coord_start\"])\n",
    "\n",
    "    # Combina todas as condições\n",
    "    df[\"condicao_valida\"] = condicao_c1_c2_positivo | condicao_c2_c1_positivo | condicao_c1_c2_negativo | condicao_c2_c1_negativo\n",
    "\n",
    "    # Identifica linhas inválidas\n",
    "    linhas_invalidas = df[~df[\"condicao_valida\"]]\n",
    "    print(\"Linhas que violam a teoria:\")\n",
    "    print(linhas_invalidas[[\"reference_exon_coord\", \"spliced_with_coord\", \"junction_coord\", \"junction_name\", \"strand\"]])\n",
    "\n",
    "    # Estatísticas\n",
    "    print(\"\\nContagem de valores em 'condicao_valida':\")\n",
    "    print(df[\"condicao_valida\"].value_counts())\n",
    "\n",
    "# Aplicar a função ao DataFrame\n",
    "check_coordinate_conditions(tandem_cassette_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mutually exclusive exons\n",
    "Mutually exclusive exons are a form of alternative splicing where only one exon from a pair (or set) of exons is included in the mature mRNA transcript, while the other exon(s) are excluded. This mechanism allows for the generation of different protein isoforms from a single gene, contributing to protein diversity.\n",
    "\n",
    "- When 'strand' is \"+\", 'reference_exon_coord_start' should be less than 'spliced_with_coord_start'.\n",
    "\n",
    "- When 'strand' is \"-\", the opposite is true: 'reference_exon_coord_start' should be greater than 'spliced_with_coord_start'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                   module_id                  gene_id  gene_name seqid strand  \\\n",
      "0  AT1G10600.Araport11.447_1  AT1G10600.Araport11.447  AT1G10600  Chr1      +   \n",
      "1  AT1G10600.Araport11.447_1  AT1G10600.Araport11.447  AT1G10600  Chr1      +   \n",
      "2  AT1G10600.Araport11.447_1  AT1G10600.Araport11.447  AT1G10600  Chr1      +   \n",
      "3  AT1G10600.Araport11.447_1  AT1G10600.Araport11.447  AT1G10600  Chr1      +   \n",
      "4  AT5G42770.Araport11.447_1  AT5G42770.Araport11.447  AT5G42770  Chr5      +   \n",
      "\n",
      "                                        lsv_id  \\\n",
      "0    AT1G10600.Araport11.447:s:3504024-3504113   \n",
      "1    AT1G10600.Araport11.447:s:3504024-3504113   \n",
      "2    AT1G10600.Araport11.447:t:3504569-3504643   \n",
      "3    AT1G10600.Araport11.447:t:3504569-3504643   \n",
      "4  AT5G42770.Araport11.447:s:17152530-17152604   \n",
      "\n",
      "                          event_id  complex  denovo reference_exon_coord  ...  \\\n",
      "0  AT1G10600.Araport11.447_1_mxe_1    False   False      3504024-3504113  ...   \n",
      "1  AT1G10600.Araport11.447_1_mxe_1    False   False      3504024-3504113  ...   \n",
      "2  AT1G10600.Araport11.447_1_mxe_1    False   False      3504569-3504643  ...   \n",
      "3  AT1G10600.Araport11.447_1_mxe_1    False   False      3504569-3504643  ...   \n",
      "4  AT5G42770.Araport11.447_1_mxe_1    False    True    17152530-17152604  ...   \n",
      "\n",
      "  event_changing junction_changing SRR11684912_median_psi SRR11684912_var_psi  \\\n",
      "0            NaN               NaN                0.05805            0.001352   \n",
      "1            NaN               NaN                0.94190            0.001352   \n",
      "2            NaN               NaN                0.93700            0.001445   \n",
      "3            NaN               NaN                0.06296            0.001445   \n",
      "4            NaN               NaN                0.30360            0.015880   \n",
      "\n",
      "   reference_exon_coord_start  reference_exon_coord_end  \\\n",
      "0                     3504024                   3504113   \n",
      "1                     3504024                   3504113   \n",
      "2                     3504569                   3504643   \n",
      "3                     3504569                   3504643   \n",
      "4                    17152530                  17152604   \n",
      "\n",
      "   spliced_with_coord_start  spliced_with_coord_end  junction_coord_start  \\\n",
      "0                   3504359                 3504392               3504113   \n",
      "1                   3504426                 3504488               3504113   \n",
      "2                   3504359                 3504392               3504488   \n",
      "3                   3504426                 3504488               3504392   \n",
      "4                  17152714                17152762              17152604   \n",
      "\n",
      "   junction_coord_end  \n",
      "0             3504359  \n",
      "1             3504426  \n",
      "2             3504569  \n",
      "3             3504569  \n",
      "4            17152714  \n",
      "\n",
      "[5 rows x 25 columns]\n"
     ]
    }
   ],
   "source": [
    "mutually_exclusive_df = pd.read_csv(mutually_exclusive, sep=\"\\t\", header=0, comment=\"#\")\n",
    "\n",
    "mutually_exclusive_df[[\"reference_exon_coord_start\", \"reference_exon_coord_end\"]] = mutually_exclusive_df[\"reference_exon_coord\"].str.split(\"-\", expand=True)\n",
    "mutually_exclusive_df[[\"spliced_with_coord_start\", \"spliced_with_coord_end\"]] = mutually_exclusive_df[\"spliced_with_coord\"].str.split(\"-\", expand=True)\n",
    "mutually_exclusive_df[[\"junction_coord_start\", \"junction_coord_end\"]] = mutually_exclusive_df[\"junction_coord\"].str.split(\"-\", expand=True)\n",
    "\n",
    "# convert start and end to int\n",
    "mutually_exclusive_df[\"reference_exon_coord_start\"] = mutually_exclusive_df[\"reference_exon_coord_start\"].astype(int)\n",
    "mutually_exclusive_df[\"reference_exon_coord_end\"] = mutually_exclusive_df[\"reference_exon_coord_end\"].astype(int)\n",
    "mutually_exclusive_df[\"spliced_with_coord_start\"] = mutually_exclusive_df[\"spliced_with_coord_start\"].astype(int)\n",
    "mutually_exclusive_df[\"spliced_with_coord_end\"] = mutually_exclusive_df[\"spliced_with_coord_end\"].astype(int)\n",
    "mutually_exclusive_df[\"junction_coord_start\"] = mutually_exclusive_df[\"junction_coord_start\"].astype(int)\n",
    "mutually_exclusive_df[\"junction_coord_end\"] = mutually_exclusive_df[\"junction_coord_end\"].astype(int)\n",
    "\n",
    "print(mutually_exclusive_df.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Linhas que violam a teoria:\n",
      "Empty DataFrame\n",
      "Columns: [module_id, gene_id, gene_name, seqid, strand, lsv_id, event_id, complex, denovo, reference_exon_coord, spliced_with, spliced_with_coord, junction_name, junction_coord, event_non_changing, event_changing, junction_changing, SRR11684912_median_psi, SRR11684912_var_psi, reference_exon_coord_start, reference_exon_coord_end, spliced_with_coord_start, spliced_with_coord_end, junction_coord_start, junction_coord_end, condicao_valida]\n",
      "Index: []\n",
      "\n",
      "[0 rows x 26 columns]\n",
      "A1    14\n",
      "A2    14\n",
      "Name: spliced_with, dtype: int64 True    28\n",
      "Name: condicao_valida, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "condicao_e2_positivo = (mutually_exclusive_df[\"strand\"] == \"+\") & (mutually_exclusive_df[\"reference_exon_coord_start\"] < mutually_exclusive_df[\"spliced_with_coord_start\"])\n",
    "condicao_e2_negativo = (mutually_exclusive_df[\"strand\"] == \"-\") & (mutually_exclusive_df[\"reference_exon_coord_start\"] > mutually_exclusive_df[\"spliced_with_coord_start\"])\n",
    "condicao_e1_positivo = (mutually_exclusive_df[\"strand\"] == \"+\") & (mutually_exclusive_df[\"reference_exon_coord_start\"] > mutually_exclusive_df[\"spliced_with_coord_start\"])\n",
    "condicao_e1_negativo = (mutually_exclusive_df[\"strand\"] == \"-\") & (mutually_exclusive_df[\"reference_exon_coord_start\"] < mutually_exclusive_df[\"spliced_with_coord_start\"])\n",
    "\n",
    "mutually_exclusive_df[\"condicao_valida\"] = condicao_e2_positivo | condicao_e2_negativo | condicao_e1_positivo | condicao_e1_negativo\n",
    "linhas_invalidas = mutually_exclusive_df[~mutually_exclusive_df[\"condicao_valida\"]]\n",
    "print(\"Linhas que violam a teoria:\")\n",
    "print(linhas_invalidas)\n",
    "print(mutually_exclusive_df[\"spliced_with\"].value_counts(), mutually_exclusive_df[\"condicao_valida\"].value_counts())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "usr",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
